lighthouse-treemap/app/src/main.js

```diff
@@ -9,6 +9,253 @@
+   */
+  createRootNodeForGroup(group) {
+    const rootNodes = this.rootNodesByGroup[group];
+
+    const children = rootNodes.map(rootNode => {
+      // TODO: keep?
+      // Wrap with the name of the rootNode. Only for bundles.
+      if (group === 'scripts' && rootNode.node.children) {
+        return {
+          name: rootNode.name,
+          children: [rootNode.node],
+          resourceBytes: rootNode.node.resourceBytes,
+          unusedBytes: rootNode.node.unusedBytes,
+        };
+      }
+
+      return rootNode.node;
+    });
+
+    return {
+      name: this.documentUrl,
+      resourceBytes: children.reduce((acc, cur) => cur.resourceBytes + acc, 0),
+      unusedBytes: children.reduce((acc, cur) => (cur.unusedBytes || 0) + acc, 0),
+      children,
+    };
+  }
+
+  show() {
+    const group = 'scripts';
+    this.currentRootNode = this.createRootNodeForGroup(group);
+    const rootNodes = this.rootNodesByGroup[group];
+    renderViewModeOptions(rootNodes);
+
+    TreemapUtil.dfs(this.currentRootNode, node => {
+      // @ts-ignore: webtreemap will store `dom` on the data to speed up operations.
+      // However, when we change the underlying data representation, we need to delete
+      // all the cached DOM elements. Otherwise, the rendering will be incorrect when,
+      // for example, switching between "All JavaScript" and a specific bundle.
+      delete node.dom;
+
+      // @ts-ignore: webtreemap uses `size` to partition the treemap.
+      node.size = node.resourceBytes || 0;
+    });
+    webtreemap.sort(this.currentRootNode);
+
+    this.el.innerHTML = '';
+    this.render();
+  }
+
+  render() {
+    this.treemap = new webtreemap.TreeMap(this.currentRootNode, {
+      padding: [18, 3, 3, 3],
+      spacing: 10,
+      caption: node => this.makeCaption(node),
+      // showChildren: node => node.children && node.children.some(c => c.resourceBytes > 1000 * 100),
+      // showNode: node => node.resourceBytes > 100 * 100,
+      // lowerBound: 0.2,
+    });
+    this.treemap.render(this.el);
+    TreemapUtil.find('.webtreemap-node').classList.add('webtreemap-node--root');
+    this.updateColors();
+  }
+
+  resize() {
+    if (!this.treemap) throw new Error('must call .render() first');
+
+    this.treemap.layout(this.currentRootNode, this.el);
+    this.updateColors();
+  }
+
+  /**
+   * Creates the header text for each node in webtreemap.
+   * @param {LH.Treemap.Node} node
+   */
+  makeCaption(node) {
+    const size = node.resourceBytes;
+    const total = this.currentRootNode.resourceBytes;
+
+    const parts = [
+      TreemapUtil.elide(node.name, 60),
+      `${TreemapUtil.formatBytes(size)} (${Math.round(size / total * 100)}%)`,
+    ];
+
+    // Only add label for bytes on the root node.
+    if (node === this.currentRootNode) {
+      parts[1] = `resource bytes: ${parts[1]}`;
+    }
+
+    return parts.join(' · ');
+  }
+
+  updateColors() {
+    TreemapUtil.dfs(this.currentRootNode, node => {
```

Comment:
figured out a decent little hack.

```diff
diff --git a/lighthouse-treemap/app/src/main.js b/lighthouse-treemap/app/src/main.js
index f3c41fe12..e2cc80221 100644
--- a/lighthouse-treemap/app/src/main.js
+++ b/lighthouse-treemap/app/src/main.js
@@ -195,6 +195,8 @@ class TreemapViewer {
     TreemapUtil.dfs(this.currentRootNode, node => {
       // Color a root node and all children the same color.
       const rootNode = this.nodeToRootNodeMap.get(node);
+      if (rootNode) return;
+
       const hueKey = rootNode ? rootNode.name : node.name;
       const hue = this.getHue(hueKey);
 
@@ -213,7 +215,7 @@ class TreemapViewer {
       // @ts-ignore: webtreemap will add a dom node property to every node.
       const dom = /** @type {HTMLElement?} */ (node.dom);
       if (dom) {
-        dom.style.backgroundColor = backgroundColor;
+        dom.style.setProperty('--node-bg-color', backgroundColor);
         dom.style.color = color;
       }
     });
diff --git a/lighthouse-treemap/app/styles/treemap.css b/lighthouse-treemap/app/styles/treemap.css
index 2dd987be0..ec739279c 100644
--- a/lighthouse-treemap/app/styles/treemap.css
+++ b/lighthouse-treemap/app/styles/treemap.css
@@ -72,6 +73,7 @@ header {
   overflow: hidden;
   background: white;
   transition: left .15s, top .15s, width .15s, height .15s;
+  background-color: var(--node-bg-color);
 }
```

i think this also means you can drop the `click` listener on L67 that's only there to run `updateColors()`

Reply:
This might not work for coloring in other modes (as it relies on the fact that for 'scripts', nodes not found in that map are derived from the data and 1:1 with unique scripts). let's wait for more modes before optimizing.

Reply:
can you add a TODO comment or item in your tracking issue for this?

another option is adding a hook in webtreemap where it'd make sense to run `updateNodeColor` instead of the 3 separate invocations.